File: ClassEval_T/cpp/solution/ClassRegistrationSystem.cpp
#include <string>
#include <vector>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>

class Student {
public:
    Student(const std::string& name, const std::string& major);

    std::string get_name() const;
    std::string get_major() const;

    void set_name(const std::string& name);
    void set_major(const std::string& major);

    bool operator==(const Student& other) const;

    std::string name;
    std::string major;
};

class ClassRegistrationSystem {
public:
    int register_student(const Student new_student);
    std::vector<std::string> register_class(const std::string& student_name, const std::string& class_name);
    std::vector<std::string> get_students_by_major(const std::string& major) const;
    std::vector<std::string> get_all_major() const;
    std::string get_most_popular_class_in_major(const std::string& major) const;
    std::vector<Student> students;
    std::unordered_map<std::string, std::vector<std::string>> students_registration_classes;
};


Student::Student(const std::string& name, const std::string& major) : name(name), major(major) {}

std::string Student::get_name() const {
    return name;
}

std::string Student::get_major() const {
    return major;
}

void Student::set_name(const std::string& name) {
    this->name = name;
}

void Student::set_major(const std::string& major) {
    this->major = major;
}

bool Student::operator==(const Student& other) const {
    return name == other.name && major == other.major;
}


int ClassRegistrationSystem::register_student(const Student new_student) {
    
    std::string name = new_student.get_name();
    std::string major = new_student.get_major();
    for (const auto& student : students) {
        if (student.get_name() == name) {
            return 0; 
        }
    }
    students.emplace_back(name, major);
    return 1; 
}

std::vector<std::string> ClassRegistrationSystem::register_class(const std::string& student_name, const std::string& class_name) {
    auto& classes = students_registration_classes[student_name];
    if (std::find(classes.begin(), classes.end(), class_name) == classes.end()) {
        classes.push_back(class_name);
    }
    return classes;
}

std::vector<std::string> ClassRegistrationSystem::get_students_by_major(const std::string& major) const {
    std::vector<std::string> student_list;
    for (const auto& student : students) {
        if (student.get_major() == major) {
            student_list.push_back(student.get_name());
        }
    }
    return student_list;
}

std::vector<std::string> ClassRegistrationSystem::get_all_major() const {
    std::set<std::string> majors;
    for (const auto& student : students) {
        
        majors.insert(student.get_major());
    }
    return std::vector<std::string>(majors.begin(), majors.end());
}

std::string ClassRegistrationSystem::get_most_popular_class_in_major(const std::string& major) const {
    std::unordered_map<std::string, int> class_count;

    
    for (const auto& student : students) {
        if (student.get_major() == major) {
            auto it = students_registration_classes.find(student.get_name());
            if (it != students_registration_classes.end()) {
                for (const auto& class_name : it->second) {
                    class_count[class_name]++;
                }
            }
        }
    }

    
    std::string most_popular_class;
    int max_count = 0;
    for (const auto& [class_name, count] : class_count) {
        if (count > max_count) {
            most_popular_class = class_name;
            max_count = count;
        }
    }

    return most_popular_class;
}

File: BatchTestTool/java/java automated tester/Test/compile_passed/solutions/ClassRegistrationSystem.java
import java.util.*;

class Student {
    private String name;
    private String major;

    public Student(String name, String major) {
        this.name = name;
        this.major = major;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getMajor() {
        return major;
    }

    public void setMajor(String major) {
        this.major = major;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Student student = (Student) o;
        return Objects.equals(name, student.name) &&
               Objects.equals(major, student.major);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, major);
    }
}

class ClassRegistrationSystem {
    private List<Student> students;
    private Map<String, List<String>> studentsRegistrationClasses;

    public ClassRegistrationSystem() {
        students = new ArrayList<>();
        studentsRegistrationClasses = new HashMap<>();
    }

    public int registerStudent(Student newStudent) {
        String name = newStudent.getName();
        String major = newStudent.getMajor();

        for (Student student : students) {
            if (student.getName().equals(name)) {
                return 0; // Student already exists
            }
        }

        students.add(new Student(name, major));
        return 1; // New student added
    }

    public List<String> registerClass(String studentName, String className) {
        List<String> classes = studentsRegistrationClasses.computeIfAbsent(studentName, k -> new ArrayList<>());

        if (!classes.contains(className)) {
            classes.add(className);
        }

        return classes;
    }

    public List<String> getStudentsByMajor(String major) {
        ArrayList<String> studentList = new ArrayList<>();
        for (Student student : students) {
            if (student.getMajor().equals(major)) {
                studentList.add(student.getName());
            }
        }
        return studentList;
    }

    public List<String> getAllMajor() {
        Set<String> majors = new HashSet<>();
        for (Student student : students) {
            majors.add(student.getMajor());
        }
        return new ArrayList<>(majors);
    }

    public String getMostPopularClassInMajor(String major) {
        Map<String, Integer> classCount = new HashMap<>();

        for (Student student : students) {
            if (student.getMajor().equals(major)) {
                List<String> classes = studentsRegistrationClasses.get(student.getName());
                if (classes != null) {
                    for (String className : classes) {
                        classCount.put(className,
                            classCount.getOrDefault(className, 0) + 1);
                    }
                }
            }
        }

        String mostPopularClass = "";
        int maxCount = 0;

        for (Map.Entry<String, Integer> entry : classCount.entrySet()) {
            if (entry.getValue() > maxCount) {
                mostPopularClass = entry.getKey();
                maxCount = entry.getValue();
            }
        }

        return mostPopularClass;
    }

    public List<Student> getStudents() {
        return students;
    }

    public Map<String, List<String>> getStudentsRegistrationClasses() {
        return studentsRegistrationClasses;
    }
}